pkg/steps/multi_stage.go

```diff
@@ -0,0 +1,202 @@
+package steps
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"strings"
+
+	coreapi "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	meta "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	coreclientset "k8s.io/client-go/kubernetes/typed/core/v1"
+
+	"github.com/openshift/ci-tools/pkg/api"
+	"github.com/openshift/ci-tools/pkg/junit"
+)
+
+const (
+	multiStageTestLabel = "ci.openshift.io/multi-stage-test"
+)
+
+type multiStageTestStep struct {
+	name            string
+	config          *api.ReleaseBuildConfiguration
+	podClient       PodClient
+	jobSpec         *api.JobSpec
+	pre, test, post []api.TestStep
+	subTests        []*junit.TestCase
+}
+
+func MultiStageTestStep(
+	testConfig api.TestStepConfiguration,
+	config *api.ReleaseBuildConfiguration,
+	podClient PodClient,
+	jobSpec *api.JobSpec,
+) (api.Step, error) {
+	return newMultiStageTestStep(testConfig, config, podClient, jobSpec), nil
+}
+
+func newMultiStageTestStep(
+	testConfig api.TestStepConfiguration,
+	config *api.ReleaseBuildConfiguration,
+	podClient PodClient,
+	jobSpec *api.JobSpec,
+) *multiStageTestStep {
+	return &multiStageTestStep{
+		name:      testConfig.As,
+		config:    config,
+		podClient: podClient,
+		jobSpec:   jobSpec,
+		pre:       testConfig.MultiStageTestConfiguration.Pre,
+		test:      testConfig.MultiStageTestConfiguration.Test,
+		post:      testConfig.MultiStageTestConfiguration.Post,
+	}
+}
+
+func (s *multiStageTestStep) Inputs(ctx context.Context, dry bool) (api.InputDefinition, error) {
+	return nil, nil
+}
+
+func (s *multiStageTestStep) Run(ctx context.Context, dry bool) error {
+	if dry {
+		return nil
```

Comment:
The other steps output a JSON of what would they submit while in dry mode, I think we should do the same.

Reply:
Yep, we will need to parse all three stages.

Reply:
It's more complex here, but I think that's doable.